include/unifex/async_scope.hpp

```diff
@@ -632,76 +625,129 @@ class _attached_op<Sender, Receiver>::type final {
   using cleaning_receiver_t = cleaning_receiver<type, Receiver>;
   using nested_operation_t = connect_result_t<Sender, cleaning_receiver_t>;
 
+  struct stop_callback {
+    void operator()() noexcept { op_.request_stop(); }
+
+    type& op_;
+  };
+
 public:
   template <typename Receiver2>
   explicit type(Sender&& s, Receiver2&& r, async_scope* scope) noexcept(
       is_nothrow_connectable_v<Sender, cleaning_receiver_t>&&
-          std::is_nothrow_constructible_v<cleaning_receiver_t, type*, Receiver2>&&
+          std::is_nothrow_constructible_v<cleaning_receiver_t, type*>&&
               std::is_nothrow_constructible_v<Receiver, Receiver2>)
-    : scope_(reinterpret_cast<std::uintptr_t>(scope) & started_mask)
-    , receiverToken_(get_stop_token(r)) {
-    if (scope_ref()) {
-      unifex::activate_union_member_with(op_, [&] {
-        return unifex::connect(
-            (Sender &&) s, cleaning_receiver_t{this, (Receiver2 &&) r});
+    : scope_(set_refcount(scope))
+    , receiver_((Receiver2 &&) r) {
+    if (scope) {
+      op_.construct_with([&, this] {
+        return unifex::connect((Sender &&) s, cleaning_receiver_t{this});
       });
-    } else {
-      unifex::activate_union_member(receiver_, (Receiver2 &&) r);
     }
   }
 
   type(type&& op) = delete;
 
   ~type() {
-    auto scope = scope_ref();
-    if (scope) {
-      unifex::deactivate_union_member(op_);
+    auto scope = scope_.load(std::memory_order_relaxed);
+    if (scope != 0u) {
+      op_.destruct();
       // started => receiver responsible for cleanup
-      if (!has_started()) {
-        record_done(scope);
+      if ((scope & ~mask) != 0u) {
+        UNIFEX_ASSERT((scope & ~mask) == 1u);
+        record_done(reinterpret_cast<async_scope*>(scope & mask));
```

Comment:
I think some of the bit-bashing code would be clearer with a couple of static methods:
```
static async_scope* scope_ptr(std::uintptr_t scope) noexcept {
  return reinterpret_cast<async_scope*>(scope & mask);
}

static std::uintptr_t ref_count(std::uintptr_t scope) noexcept {
  return scope & ~mask;
}
```

Then you could write this code thus:
```suggestion
      if (ref_count(scope) != 0u) {
        UNIFEX_ASSERT(ref_count(scope) == 1u);
        record_done(scope_ptr(scope));
```